jsregexp.c

```diff
@@ -262,10 +269,172 @@ static int regexp_tostring(lua_State *lstate)
 }
 
 
+// automatic conversion to the global match string
+static int match_tostring(lua_State *lstate)
+{
+  //luaL_getmetatable(lstate, JSREGEXP_MATCH);
+  //if (!lua_getmetatable(lstate, 1) || !lua_equal(lstate, -1, -2)) {
+  //  luaL_argerror(lstate, 1, "match object expected");
+  //}
+  lua_rawgeti(lstate, 1, 0);
+  return 1;
+}
+
+
+// repeatedly running regexp:match(str) is not a good idea because we would
+// convert the string (at least from last_ind) to utf16 every time (if it is
+// needed)
+static int regexp_exec(lua_State *lstate)
+{
+  uint8_t *capture[CAPTURE_COUNT_MAX * 2];
+
+  size_t input_len;
+  struct regexp *r = luaL_checkudata(lstate, 1, JSREGEXP_MT);
+  const char *input = luaL_checklstring(lstate, 2, &input_len);
+
+  // check what happens in JS
+  if (r->last_index >= input_len) {
+    r->last_index = input_len;
+  }
+
+  const int capture_count = lre_get_capture_count(r->bc);
+  const int global = lre_get_flags(r->bc) & LRE_FLAG_GLOBAL;
+  const char* group_names = lre_get_groupnames(r->bc);
+
+  const bool matched = lre_exec(capture, r->bc, (uint8_t *) input,
```

Comment:
This is actually a three state value (1=match, 0=no match, -1 = error), it can fail by running out of memory.

Reply:
Yea, I am only checking against `1`, is there anything we can do on error, realistically?

Reply:
Just throw an error saying out of memory. An error is better than silently failing and giving an invalid result. It is also what quickjs does